### ISSUE TITLE
[API Proposal]: Support of configuration options with a full qualified section key like in the old days of .NET Framework

### ISSUE THREAD

**Issue by HHobeck** [dotnet/runtime#95678]
### Background and motivation

Dear community,

I came up with the idea to extend the Microsoft.Extensions.Options model to support the configuration options with a full qualified section key to resolves the IOptions<T> dependencies automatically. This has the advantage that you don't need to register the configuration classes explicit in the application startup each time. It is useful especially if you are not aware of third party library and wants to change the behavior or fine tune the settings (via IOptions<T>) in a later stage without recompilation.

Hope this finds a way in .net core because it was a breaking change compared to the old world of .NET Framework and left the user without the option to change the behavior of how the settings are resolved.

Thank you.

### API Proposal

Possible implementation:
```csharp
using System;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;

namespace Microsoft.Extensions.Options;

public static class OptionsWithFullQualifiedSectionKeyComposition
{
    public static void Register(IServiceCollection serviceCollection)
    {
        if (serviceCollection is null) throw new ArgumentNullException(nameof(serviceCollection));

        serviceCollection.AddSingleton(typeof(IConfigureOptions<>), typeof(ConfigureOptionsWithFullQualifiedSectionKey<>));
        serviceCollection.AddSingleton(
            typeof(IOptionsChangeTokenSource<>), typeof(ConfigurationChangeTokenSourceWithFullQualifiedSectionKey<>)
        );
    }
}

public sealed class ConfigureOptionsWithFullQualifiedSectionKey<TOptions>
    : ConfigureFromConfigurationOptions<TOptions> where TOptions : class
{
    public ConfigureOptionsWithFullQualifiedSectionKey(IConfiguration configuration)
        : base(configuration?.GetSection(typeof(TOptions).FullName)!)
    {
    }
}

public sealed class ConfigurationChangeTokenSourceWithFullQualifiedSectionKey<TOptions>
    : ConfigurationChangeTokenSource<TOptions>
{
    public ConfigurationChangeTokenSourceWithFullQualifiedSectionKey(IConfiguration configuration)
        : base(configuration?.GetSection(typeof(TOptions).FullName)!)
    {
    }
}
```

### API Usage

Possible usage:
```csharp
internal class Program
{
    static void Main(string[] arguments)
    {
        IHostBuilder hostBuilder = Host.CreateDefaultBuilder(arguments);
        hostBuilder.ConfigureServices(
            serviceCollection => OptionsWithFullQualifiedSectionKeyComposition.Register(serviceCollection)
        );

        using IHost host = hostBuilder.Build();
        host.Run();
    }
}
```

### Risks

This has no risks and can be seen as an option to the user to change the way how the configuration are loaded.

**Comment by HHobeck** [comment]
@tarekgh Maybe this feature is something for .NET 10? ;)

**Comment by tarekgh** [comment]
@HHobeck Thanks for reaching out. I’m monitoring this request to assess demand before considering support for it. Does this pose a blocker for you in any way? It looks to me trivial to add this code to your app if necessary.

**Comment by HHobeck** [comment]
Hi @tarekgh. 
No it's not a blocker for me. I' using it in production code for years already. My motivation just is to bring this feature into the .NET world, because I think it is very convenient and a missing puzzle.

**Comment by HHobeck** [comment]
> The proposed implementation gets the name of the configuration section from `typeof(TOptions).FullName`. That can get unwieldy if TOptions is a constructed generic type. For example, `typeof(List<int>.FullName)` is `` System.Collections.Generic.List`1[[System.Int32, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]] ``, i.e. it includes the assembly-qualified name of the generic argument. When Microsoft.Extensions.Logging.LoggerFactoryExtensions translates types to log category names, it uses the internal class [TypeNameHelper](https://github.com/dotnet/runtime/blob/9d5a6a9aa463d6d10b0b0ba6d5982cc82f363dc3/src/libraries/Common/src/Extensions/TypeNameHelper/TypeNameHelper.cs) to format the type names more concisely.

Yes, that's a good point. Thank you for bringing this to our attention.
